### PR TITLE
Backward process in recurrent layers

### PR DESCRIPTION
**Pull request by dbonadiman**
After weeks of testing and actively using this code i decided to ask for a pull request.

Basically i added the option to process the input sequence in backward mode when using Recurrent layers.
This is an add only option since by default the sequence is processed forward.

Adding this possibility however makes possible to implement a Bidirectional LSTM using a Graph model.

In that sense i added an example that make use of this new API and finally the (so hated) Graph model.

### CODE REVIEW

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -141,7 +143,7 @@ def __init__(self, output_dim, depth=3,
                  init='glorot_uniform', inner_init='orthogonal',
                  activation='sigmoid', inner_activation='hard_sigmoid',
                  weights=None, truncate_gradient=-1, return_sequences=False,
-                 input_dim=None, input_length=None, **kwargs):
+                 input_dim=None, input_length=None, backwards=False,**kwargs):
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Space between arguments (PEP8).

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ class SimpleRNN(Recurrent):
     '''
     def __init__(self, output_dim,
                  init='glorot_uniform', inner_init='orthogonal', activation='sigmoid', weights=None,
-                 truncate_gradient=-1, return_sequences=False, input_dim=None, input_length=None, **kwargs):
+                 truncate_gradient=-1, return_sequences=False, input_dim=None, input_length=None, backwards=False, **kwargs):
```
Comment (fchollet) [inline keras/layers/recurrent.py]: Why `backwards` argument instead of `go_backwards`?

Comment (fchollet) [inline keras/layers/recurrent.py]: I mean, it's a pretty serious inconsistency.

File: examples/imdb_bidirectional_lstm.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+from __future__ import absolute_import
+from __future__ import print_function
+import numpy as np
+np.random.seed(1337)  # for reproducibility
+
+from keras.preprocessing import sequence
+from keras.optimizers import SGD, RMSprop, Adagrad
```
Comment (fchollet) [inline examples/imdb_bidirectional_lstm.py]: Lots of imports that you aren't actually using.

File: examples/imdb_bidirectional_lstm.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+from keras.utils import np_utils
+from keras.models import Sequential, Graph
+from keras.layers.core import Dense, Dropout
+from keras.layers.embeddings import Embedding
+from keras.layers.recurrent import LSTM, GRU
+from keras.datasets import imdb
+
+from sklearn.metrics import accuracy_score as accuracy
```
Comment (fchollet) [inline examples/imdb_bidirectional_lstm.py]: Please do not import anything from sklearn, as it isn't supposed to be a Keras depency. We have our own accuracy utils in `keras/utils/np_utils.py`.

File: examples/imdb_bidirectional_lstm.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    The dataset is actually too small for LSTM to be of any advantage
+    compared to simpler, much faster methods such as TF-IDF+LogReg.
+    Notes:
+    - RNNs are tricky. Choice of batch size is important,
+    choice of loss and optimizer is critical, etc.
+    Some configurations won't converge.
+    - LSTM loss decrease patterns during training can be quite different
+    from what you see with CNNs/MLPs/etc.
+    GPU command:
+        THEANO_FLAGS=mode=FAST_RUN,device=gpu,floatX=float32 python imdb_lstm.py
```
Comment (fchollet) [inline examples/imdb_bidirectional_lstm.py]: A description that is customized to this specific example would be preferable, specifically point out what there is to know about bidirectional networks and what should be expected of them.

File: examples/imdb_bidirectional_lstm.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+y_test = np.array(y_test)
+
+print('Build model...')
+model = Graph()
+model.add_input(name='input', input_shape=(1,), dtype=int)
+model.add_node(Embedding(max_features, 128, input_length=maxlen), name='embedding', input='input')
+model.add_node(LSTM(64), name='forward', input='embedding' )  # You can change these two layers with GRU
+model.add_node(LSTM(64, backwards=True), name='backward', input='embedding' )
+model.add_node(Dropout(0.5), name='drop', inputs=['forward', 'backward'])
+model.add_node(Dense(1, activation='sigmoid'), name='out_layer', input='drop')
```
Comment (fchollet) [inline examples/imdb_bidirectional_lstm.py]: Why not `dropout`, `output`? Would seem nicer.

File: examples/imdb_bidirectional_lstm.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+model.add_output(name='output', input='out_layer')
+
+# try using different optimizers and different optimizer configs
+model.compile('adam',{'output':'binary_crossentropy'})
+
+print("Train...")
+model.fit({'input':X_train, 'output':y_train}, batch_size=batch_size, nb_epoch=4)
+acc = accuracy(y_test, np.round(np.array(model.predict({'input':X_test}, batch_size=batch_size)['output'])))
+print('Test score:', acc)
```
Comment (fchollet) [inline examples/imdb_bidirectional_lstm.py]: "test accuracy"

**Comment by fchollet** [comment]
Looking good. We can merge this after you address my comments. Also, your commits will need to be squashed into a single commit to preserve commit history readability.

**Comment by dbonadiman** [comment]
@fchollet  I close this pull request in favour of a nicer one.
I will do a clean fork and add these modification on the up to date keras, master.
